File: 2023/src/bin/day1/main.rs
#![warn(clippy::pedantic)]
use std::borrow::Cow;

use advent_2023::{BasicSolution, Solution};
use anyhow::anyhow;

struct Day;

impl BasicSolution for Day {
    const INPUT: &'static str = include_str!("data.txt");
    const SAMPLE_INPUT: &'static str = include_str!("sample_a.txt");
    const SAMPLE_INPUT_B: &'static str = include_str!("sample_b.txt");

    type Common = &'static str;
    type Answer = u32;

    const SAMPLE_ANSWER_A: Self::TestAnswer = 142;
    const SAMPLE_ANSWER_B: Self::TestAnswer = 281;

    fn common(input: &'static str) -> anyhow::Result<Self::Common> {
        Ok(input)
    }

    fn part_a(document: Cow<Self::Common>) -> anyhow::Result<Self::Answer> {
        solve(&document, &[])
    }

    fn part_b(document: Self::Common) -> anyhow::Result<Self::Answer> {
        solve(
            document,
            &[
                ("one", 1),
                ("two", 2),
                ("three", 3),
                ("four", 4),
                ("five", 5),
                ("six", 6),
                ("seven", 7),
                ("eight", 8),
                ("nine", 9),
            ],
        )
    }
}

fn solve(document: &str, spelled_out_vals: &[(&str, u32)]) -> anyhow::Result<u32> {
    let calibration_value = |line: &str| {
        let err = || anyhow!("Couldn't find a digit in line '{line}'");

        let digit_at_i = |i| {
            let literal = line[i..=i].parse().ok();
            let match_spelled = |&(digit, val)| line[i..].starts_with(digit).then_some(val);
            literal.or_else(|| spelled_out_vals.iter().find_map(match_spelled))
        };

        let first = (0..line.len()).find_map(digit_at_i).ok_or_else(err)?;
        let last = (0..line.len()).rev().find_map(digit_at_i).ok_or_else(err)?;

        Ok(first * 10 + last)
    };
    itertools::process_results(document.lines().map(calibration_value), |it| it.sum())
}

fn main() -> anyhow::Result<()> {
    Day::main()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a() -> anyhow::Result<()> {
        Day::test_part_a()
    }

    #[test]
    fn b() -> anyhow::Result<()> {
        Day::test_part_b()
    }
}

File: 2023/python/day1/main.py
from pathlib import Path

SAMPLE_INPUT = """1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
"""
SAMPLE_INPUT_B = """two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
"""
INPUT = Path("data.txt").read_text()


def part_a(input: str):
    return sum(extract_number(line) for line in input.splitlines())


def part_b(input: str):
    return sum(extract_number(fix_line(line)) for line in input.splitlines())


def fix_line(line: str):
    return (
        line.replace("one", "o1e")
        .replace("two", "t2o")
        .replace("three", "t3e")
        .replace("four", "f4r")
        .replace("five", "f5e")
        .replace("six", "s6x")
        .replace("seven", "s7n")
        .replace("eight", "e8t")
        .replace("nine", "n9e")
    )


def extract_number(line: str):
    first = ""
    last = ""
    for char in line:
        if char.isdigit():
            last = char
            if not first:
                first = char
    return int(first + last)


def test_part_a(capsys):
    assert part_a(SAMPLE_INPUT) == 142
    with capsys.disabled():
        print(f"Part A: {part_a(INPUT)}")


def test_part_b(capsys):
    assert part_b(SAMPLE_INPUT_B) == 281
    with capsys.disabled():
        print(f"Part B: {part_b(INPUT)}")
